File: algorithms/mergeSort.js
// Merge Sort Algorithm
// Merge sort breaks the array till they are individual elements and then merges them together in a sorted order.

// Time complexity : O (n * log n)

// Merge function which merges the array elements together.
const merge = ({ array, low, mid, high }) => {

    // Index for first (left) part of array
    let i = low
    // Index for second (right) part of array 
    let j = mid + 1

    // Index for second Array
    let k = low

    // Array in which elements will be added
    let secondArray = []

    // While both arrays are not completely traversed
    while (i <= mid && j <= high) {

        // If element in left array is smaller
        if (array[i] < array[j]) {
            // Add element in the array and increment indexes
            secondArray[k] = array[i]
            i++
            k++
        }
        // If element in right array is smaller
        else {
            // Add element in the array and increment indexes
            secondArray[k] = array[j]
            j++
            k++
        }
    }

    // If left array was not completely traversed, add all elements in the second array.
    while (i <= mid) {
        secondArray[k] = array[i]
        i++
        k++
    }

    // If right array was not completely traversed, add all elements in the second array.
    while (j <= high) {
        secondArray[k] = array[j]
        j++
        k++
    }

    // Mapping the elements to the original array (Mind the index!)
    for (let index = low; index <= high; index++) {
        array[index] = secondArray[index]
    }
}

// Mergesort function which calls itself recursively to divide the array to be sorted.
const mergeSort = ({ array, low, high }) => {
    if (low < high) {
        // Find the middle index of the array
        let mid = Math.floor((low + high) / 2)
        // Use mergeSort on elements to the left of the array
        mergeSort({ array: array, low: low, high: mid })
        // Use mergeSort on elements to the right of the array
        mergeSort({ array: array, low: mid + 1, high: high })
        // Merge the arrays
        merge({ array: array, low: low, mid: mid, high: high })
    }
}

const arr = [3, 4, 8, 4, 9, 7, 5, 6, 1, 2]

mergeSort({ array: arr, low: 0, high: arr.length - 1 })

console.log(arr)


// How it is being sorted : ⬇️

// [
//     3, 4, 8, 4, 9,
//     7, 5, 6, 1, 2
// ]
// [
//     3, 4, 8, 4, 9,
//     7, 5, 6, 1, 2
// ]
// [
//     3, 4, 8, 4, 9,
//     7, 5, 6, 1, 2
// ]
// [
//     3, 4, 4, 8, 9,
//     7, 5, 6, 1, 2
// ]
// [
//     3, 4, 4, 8, 9,
//     5, 7, 6, 1, 2
// ]
// [
//     3, 4, 4, 8, 9,
//     5, 6, 7, 1, 2
// ]
// [
//     3, 4, 4, 8, 9,
//     5, 6, 7, 1, 2
// ]
// [
//     3, 4, 4, 8, 9,
//     1, 2, 5, 6, 7
// ]
// [
//     1, 2, 3, 4, 4,
//     5, 6, 7, 8, 9
// ]
// [
//     1, 2, 3, 4, 4,
//     5, 6, 7, 8, 9
// ]

File: Java/Sorting/MergeSort.java
public class MergeSort {

    static void merge(int[] arr, int left, int mid, int right) {

        int leftSize = mid - left + 1;
        int rightSize = right - mid;

        int[] leftArr = new int[leftSize];
        int[] rightArr = new int[rightSize];

        for (int i = 0; i < leftSize; i++)
            leftArr[i] = arr[left + i];

        for (int j = 0; j < rightSize; j++)
            rightArr[j] = arr[mid + 1 + j];

        int i = 0, j = 0, k = left;

        while (i < leftSize && j < rightSize) {

            if (leftArr[i] <= rightArr[j]) {
                arr[k] = leftArr[i];
                i++;
            } else {
                arr[k] = rightArr[j];
                j++;
            }

            k++;
        }

        while (i < leftSize) {
            arr[k++] = leftArr[i++];
        }

        while (j < rightSize) {
            arr[k++] = rightArr[j++];
        }
    }

    static void mergeSort(int[] arr, int left, int right) {

        if (left >= right) {
            return;
        }

        int mid = left + (right - left) / 2;

        mergeSort(arr, left, mid);
        mergeSort(arr, mid + 1, right);

        merge(arr, left, mid, right);
    }

    public static void main(String[] args) {
        int[] arr = {38, 27, 43, 3, 9, 82, 10};

        System.out.println("Original array : ");
        for(int i:arr){
            System.out.print(i+" ");
        }

        mergeSort(arr, 0, arr.length - 1);

        System.out.println("\nSorted array : ");
        
        for(int i:arr){
            System.out.print(i+" ");
        }
    }
}
